File: backend/app/services/excel_service.py
"""Excel parsing — flexible header mapping → canonical CRM fields (port of excelService/columnMapper.js)."""
import io
from typing import Any, Dict, List, Optional

from openpyxl import load_workbook

ALIASES = {
    "brandName": ["brand name", "brand", "company", "company name", "name", "account", "account name", "organisation", "organization"],
    "brandId": ["brand id", "brandid", "id", "brand code", "account id", "crm id", "company id"],
    "owner": ["owner", "poc", "poc owner", "sales owner", "assigned to", "rep", "sales rep", "account owner"],
    "agency": ["agency", "agency name", "media agency", "ad agency"],
    "duplicateStatus": ["duplicate status", "duplicate", "is duplicate", "dup", "dup status", "duplicate flag"],
}
CANONICAL = list(ALIASES.keys())


def _norm(v) -> str:
    return " ".join(str("" if v is None else v).strip().lower().replace("_", " ").split())


def map_columns(header_row: List[Any]) -> Dict[str, Any]:
    normalized = [_norm(h) for h in (header_row or [])]
    mapping, matched, missing = {}, [], []
    for field, aliases in ALIASES.items():
        idx = -1
        for alias in aliases:
            if alias in normalized:
                idx = normalized.index(alias)
                break
        mapping[field] = idx
        (matched if idx != -1 else missing).append(field)
    return {"map": mapping, "matched": matched, "missing": missing}


def _find_header(rows: List[List[Any]]) -> Optional[Dict[str, Any]]:
    best = None
    for r in range(min(len(rows), 15)):
        m = map_columns(rows[r])
        if m["map"]["brandName"] >= 0 and (best is None or len(m["matched"]) > best["matched"]):
            best = {"index": r, "matched": len(m["matched"]), "mapping": m}
    return best


def _cell(v) -> str:
    return "" if v is None else str(v).strip()


def extract_from_rows(rows: List[List[Any]]) -> Dict[str, Any]:
    if not rows:
        return {"companies": [], "mapping": {"map": {}, "matched": [], "missing": CANONICAL}, "skipped": 0, "error": "The sheet is empty."}
    header = _find_header(rows)
    if not header:
        return {"companies": [], "mapping": map_columns(rows[0]), "skipped": 0,
                "error": "Could not find a Brand Name column in the first rows."}
    m = header["mapping"]["map"]
    companies, skipped = [], 0
    for r in range(header["index"] + 1, len(rows)):
        row = rows[r]
        if not row:
            continue
        brand = _cell(row[m["brandName"]]) if m["brandName"] < len(row) else ""
        if not brand:
            skipped += 1
            continue
        companies.append({
            "rowIndex": r + 1, "brandName": brand,
            "brandId": _cell(row[m["brandId"]]) if 0 <= m["brandId"] < len(row) else "",
            "owner": _cell(row[m["owner"]]) if 0 <= m["owner"] < len(row) else "",
            "agency": _cell(row[m["agency"]]) if 0 <= m["agency"] < len(row) else "",
            "duplicateStatus": _cell(row[m["duplicateStatus"]]) if 0 <= m["duplicateStatus"] < len(row) else "",
        })
    return {"companies": companies, "mapping": header["mapping"], "skipped": skipped}


def parse_bytes(data: bytes) -> Dict[str, Any]:
    wb = load_workbook(io.BytesIO(data), read_only=True, data_only=True)
    ws = wb[wb.sheetnames[0]]
    rows = [list(r) for r in ws.iter_rows(values_only=True)]
    result = extract_from_rows(rows)
    result["sheetName"] = wb.sheetnames[0]
    return result

File: src/services/excelService.js
/**
 * ALIP · ExcelService (service)
 * ---------------------------------------------------------------------------
 * Reads an uploaded .xlsx and extracts canonical CRM companies.
 *
 *   parseFile(file)          browser wrapper: File -> SheetJS workbook -> rows
 *   extractCompanies(rows)   PURE core: array-of-arrays -> company records
 *
 * The pure core has no dependency on the DOM or SheetJS, so it is fully unit
 * testable in Node. Identity is never fabricated: a row without a Brand Name
 * is skipped, not invented.
 *
 * Mirrors future backend service: backend/services/excel_service.py
 */
(function (root) {
  'use strict';

  var ColumnMapper = (typeof module !== 'undefined' && module.exports)
    ? require('../utils/columnMapper.js')
    : (root.ALIP && root.ALIP.ColumnMapper);

  var CANONICAL_FIELDS = ['brandName', 'brandId', 'owner', 'agency', 'duplicateStatus'];

  function cell(value) {
    return value == null ? '' : String(value).trim();
  }

  /**
   * Locate the header row. Real CRM exports often have title/blank rows above
   * the headers, so we scan the first rows and pick the one that maps the most
   * canonical fields (and must include Brand Name).
   * @returns {{ index:number, matched:number, mapping:Object }}
   */
  function findHeaderRow(rows) {
    var best = { index: -1, matched: -1, mapping: null };
    var limit = Math.min(rows.length, 15);
    for (var r = 0; r < limit; r++) {
      var mapping = ColumnMapper.mapColumns(rows[r]);
      if (mapping.map.brandName >= 0 && mapping.matched.length > best.matched) {
        best = { index: r, matched: mapping.matched.length, mapping: mapping };
      }
    }
    return best;
  }

  /**
   * Extract canonical CRM companies from a sheet represented as array-of-arrays.
   * @param {Array<Array<*>>} rows  rows[0] = header row.
   * @returns {{ companies: Object[], mapping: Object, skipped: number, error?: string }}
   */
  function extractCompanies(rows) {
    if (!rows || !rows.length) {
      return {
        companies: [],
        mapping: { map: {}, matched: [], missing: CANONICAL_FIELDS.slice() },
        skipped: 0,
        error: 'The sheet is empty.'
      };
    }

    var header = findHeaderRow(rows);
    if (header.index < 0) {
      var preview = (rows[0] || []).map(function (h) { return '"' + cell(h) + '"'; }).join(', ');
      return {
        companies: [],
        mapping: header.mapping || ColumnMapper.mapColumns(rows[0] || []),
        skipped: 0,
        error: 'Could not find a "Brand Name" column in the first rows. Headers seen: ' + preview
      };
    }

    var mapping = header.mapping;
    var map = mapping.map;
    var companies = [];
    var skipped = 0;

    for (var r = header.index + 1; r < rows.length; r++) {
      var row = rows[r];
      if (!row || !row.length) { continue; }

      var brandName = cell(row[map.brandName]);
      if (!brandName) { skipped++; continue; }   // never fabricate identity

      companies.push({
        rowIndex: r + 1,
        brandName: brandName,
        brandId:         map.brandId >= 0         ? cell(row[map.brandId])         : '',
        owner:           map.owner >= 0           ? cell(row[map.owner])           : '',
        agency:          map.agency >= 0          ? cell(row[map.agency])          : '',
        duplicateStatus: map.duplicateStatus >= 0 ? cell(row[map.duplicateStatus]) : ''
      });
    }

    return { companies: companies, mapping: mapping, skipped: skipped };
  }

  /**
   * Browser entry point. Reads a File and returns the extraction result.
   * @param {File} file
   * @returns {Promise<Object>}
   */
  function parseFile(file) {
    return new Promise(function (resolve, reject) {
      if (!file) { return reject(new Error('No file provided.')); }
      if (!/\.xlsx$/i.test(file.name)) { return reject(new Error('Please upload a .xlsx file.')); }

      var XLSX = root.XLSX;
      if (!XLSX) { return reject(new Error('Excel library (SheetJS) is not loaded.')); }

      var reader = new FileReader();
      reader.onerror = function () { reject(new Error('Could not read the file.')); };
      reader.onload = function (e) {
        try {
          var wb = XLSX.read(new Uint8Array(e.target.result), { type: 'array' });
          var sheet = wb.Sheets[wb.SheetNames[0]];
          var rows = XLSX.utils.sheet_to_json(sheet, { header: 1, blankrows: false, defval: '' });
          var result = extractCompanies(rows);
          result.fileName = file.name;
          result.sheetName = wb.SheetNames[0];
          resolve(result);
        } catch (err) {
          reject(new Error('Failed to parse workbook: ' + err.message));
        }
      };
      reader.readAsArrayBuffer(file);
    });
  }

  var ExcelService = {
    CANONICAL_FIELDS: CANONICAL_FIELDS,
    extractCompanies: extractCompanies,
    parseFile: parseFile
  };

  if (typeof module !== 'undefined' && module.exports) {
    module.exports = ExcelService;                 // Node (tests)
  } else {
    root.ALIP = root.ALIP || {};
    root.ALIP.ExcelService = ExcelService;         // Browser
  }
})(typeof self !== 'undefined' ? self : this);
